### PR TITLE
Animations with support for customization

### PR DESCRIPTION
**Pull request by naqvitalha**

### CODE REVIEW

**Comment by AbdallaMohamed** [comment]
Looks promising! What's the difference between using that `ItemAnimator` class and just rendering an `Animatable.View`?

**Comment by naqvitalha** [comment]
If you use `Animated.View` in your component you can just animate your component and not the cell RLV puts your view inside. Also, since we compute the layout we can have all kinds of fancy relayout/add/remove animations on native and web!
Simplest way to understand this is by analysing a use case where all your list items should fly from right into the list in a staggered manner.

File: src/core/RecyclerListView.tsx [hunk 1]
```diff
@@ -548,4 +556,12 @@ RecyclerListView.propTypes = {
     //outside and pass it down via this prop. Changing this object will cause everything to re-render. Make sure you don't change
     //it often to ensure performance. Re-renders are heavy.
     extendedState: PropTypes.object,
+
+    //Enables animating RecyclerListView item cells e.g, shift, add, remove etc. This prop can be used to pass an external item animation implementation.
+    //Look into BaseItemAnimator/DefaultNativeItemAnimator/DefaultWevItemAnimator for more info.
```
Comment (muskeinsingh) [inline src/core/RecyclerListView.tsx]: spell check - DefaultWevItemAnimator to DefaultWebItemAnimator

File: src/core/dependencies/DataProvider.ts [hunk 2]
```diff
@@ -1,3 +1,4 @@
+import TSCast from "../../utils/TSCast";
```
Comment (muskeinsingh) [inline src/core/dependencies/DataProvider.ts]: unused import?

File: src/platform/reactnative/DefaultNativeItemAnimator.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
+import ItemAnimator from "../../core/ItemAnimator";
+import { LayoutAnimation, Platform, UIManager } from "react-native";
+
+export default class DefaultNativeItemAnimator implements ItemAnimator {
+    public shouldAnimateOnce: boolean = true;
+    private hasAnimatedOnce: boolean = false;
+    constructor() {
+        if (Platform.OS === "android" && UIManager.setLayoutAnimationEnabledExperimental) {
```
Comment (muskeinsingh) [inline src/platform/reactnative/DefaultNativeItemAnimator.ts]: iOS?

Comment (naqvitalha) [reply src/platform/reactnative/DefaultNativeItemAnimator.ts]: Only required on android

**Comment by AbdallaMohamed** [comment]
Would it be possible to add an animation for rows that have been deleted? Where after removing a specific row, the rows below it would slide up smoothly?
Swipe to delete

**Comment by naqvitalha** [comment]
@AbdallaMohamed you can see it'll already work for resizing rows. The reason it won't work expectedly for add/delete right now is the lack of custom stableIds. That's next on the list. You will see some amazing improvements post that :)

**Comment by AbdallaMohamed** [comment]
@naqvitalha 

Thanks for the work you guys keep putting in! Can we get like a simple example of adding a fade in animation? I think this, along with the upcoming stableIds feature will be a big win over FlatList.

**Comment by AashJ** [comment]
How would we do something like animate on reorder with this approach?

**Pull request merged**